# 안녕하세요^^ 
## AIVLE 1차 미니프로젝트 '서울시 생활정보 기반 대중교통 수요 분석 ' 과정에 오신 여러분을 환영합니다.
* 본 과정에서는 실제 사례와 데이터를 기반으로 문제를 해결하는 전체 과정을 자기 주도형 실습으로 진행해볼 예정입니다.
* 앞선 교육과정을 정리하는 마음과 지금까지 배운 내용을 바탕으로 문제 해결을 해볼게요!
* 미니 프로젝트를 통한 문제 해결 과정 'A에서 Z까지', 지금부터 시작합니다!

---

# 데이터 분석부터 먼저 시작해보겠습니다.
## "구 별 유동인구 데이터" 를 확인해 보도록 하겠습니다

In [ ]:
# 필요 라이브러리부터 설치합니다.
# %pip install pandas seaborn

## 기본전제
* 처음에 제공되는 기본/추가 데이터는 '에이블러용' 폴더에 있습니다.

[기본 데이터]
* seoul_moving_month_202401.csv

[데이터 소개]
* 서울 시 구별 이동 2024년 1월 데이터
* https://data.seoul.go.kr/dataVisual/seoul/seoulLivingMigration.do

[변수 소개]
* 도착시간 : 0 -> "0:00~0:59"를 의미 
* 출발 자치구 : 자치구 코드
* 성별 : F(여성), M(남성)
* 나이 : 10 ~ 79세 (5세 단위), 0 ~ 10세, 80세 이상으로 구분
* 유형 : H(야간상주지), W(주간상주지), E(기타)
* 평균 이동 시간 : 동일 열 내 이동인구의 월 단위 평균 이동 시간 (10분단위)
* 이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리

---

# 1.데이터 불러오기
## 모든 미니 프로젝트의 시작은 '데이터 불러오기' 부터라고 할 수 있습니다.
+ KeyPoint : 불러오고자 하는 데이터에 따라 자유롭게 변수로 지정할 수 있다.

###  데이터 프레임을 불러오고 변수로 저장(여기서는 CSV 기준으로 진행)
* csv : pd.read_csv("파일이름. csv")
* txt : pd.read_csv("파일이름. csv", sep="구분자")
* xlsx : pd.read_excel('파일이름.xlsx')
* pickle : pd.read_pickle("파일이름.pkl") <br>
 [참고] pickle은 파이썬의 모든 객체를 파일로 저장할 수 있는 방법으로 DataFrame,List,Dict 등 모든 객체 저장 가능(특히 sklearn라이브러리를 통해 모델을 학습시키고, 저장할 때 많이 사용)

#### [실습문제1] 데이터 로딩
* Pandas 라이브러리를 활용해서 'seoul_moving_month_202401.csv' 데이터를 확인하세요.
    * 아래에 'Pandas'를 Import하고 'seoul_moving_month_202401.csv'파일을 'seoul_moving'에 불러오고,  Shape을 확인하는 소스코드를 작성하고 실행해보세요.
    * 데이터 파일 로딩시 참고 사항 
        * 구분자(sep)는 ',' 입니다
        * UTF-8 인코더를 사용해 주세요

In [1]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.

import pandas as pd
seoul_moving = pd.read_csv('1.2 seoul_moving_month_202401.csv',sep=",", encoding = 'UTF-8')

In [2]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 데이터 프레임의 Shape을 확인합니다.
seoul_moving.shape

(29361031, 10)

---

# 2.기본 정보 확인 및 클렌징

+ 데이터 클렌징 : 결측치, 이상치 등을 제거하여 데이터 분석 결과가 왜곡 되는 문제를 방지하기 위한 정제 과정

#### [실습문제2] 기본 정보 확인하기
* 'seoul_moving' 데이터의 정보를 확인해보세요.
* 'describe', 'info', 'head' 등 전부 활용해 보겠습니다.

In [3]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# describe()

seoul_moving.describe()

대상연월          도착시간     출발 시군구 코드     도착 시군구 코드            나이  \
count  29361031.0  2.936103e+07  2.936103e+07  2.936103e+07  2.936103e+07   
mean     202401.0  1.362194e+01  1.710538e+04  1.718123e+04  4.285979e+01   
std           0.0  5.835017e+00  9.024981e+03  9.037102e+03  1.799456e+01   
min      202401.0  0.000000e+00  1.101000e+04  1.101000e+04  0.000000e+00   
25%      202401.0  1.000000e+01  1.110000e+04  1.111000e+04  3.000000e+01   
50%      202401.0  1.400000e+01  1.120000e+04  1.120000e+04  4.000000e+01   
75%      202401.0  1.800000e+01  3.101400e+04  3.101400e+04  5.500000e+01   
max      202401.0  2.300000e+01  3.900000e+04  3.900000e+04  8.000000e+01   

        평균 이동 시간(분)  
count  2.936103e+07  
mean   9.366626e+01  
std    7.803316e+01  
min    1.000000e+00  
25%    4.900000e+01  
50%    7.400000e+01  
75%    1.120000e+02  
max    4.462000e+03

In [4]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# info()

seoul_moving.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29361031 entries, 0 to 29361030
Data columns (total 10 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   대상연월         int64 
 1   요일           object
 2   도착시간         int64 
 3   출발 시군구 코드    int64 
 4   도착 시군구 코드    int64 
 5   성별           object
 6   나이           int64 
 7   이동유형         object
 8   평균 이동 시간(분)  int64 
 9   이동인구(합)      object
dtypes: int64(6), object(4)
memory usage: 2.2+ GB


In [5]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# head()

seoul_moving.head()

대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분) 이동인구(합)
0  202401  일     0      11010      11010  F   0   EE           20   28.06
1  202401  일     0      11010      11010  F  10   EE          167   11.96
2  202401  일     0      11010      11010  F  10   EH          203   17.63
3  202401  일     0      11010      11010  F  10   HE           23    5.96
4  202401  일     0      11010      11010  F  10   WH           27    5.90

* 1번 열의 의미
    * " 2024년 1월 일요일에 11010(종로구)에서 11010(종로구)으로 이동하여 0시 ~ 0시 59분 사이에 도착했고, E(기타)에서 E(기타)로 이동한 10 ~ 15세 여성은 총 11.96명이며 이들의 평균 이동시간은 167분이다


#### [실습문제3] 결측치 처리 ( NaN이 아닌 결측치를 처리 해봅시다)

* '이동인구(합)' column은 이동 인구의 합산인데 object 타입입니다. 

    - 왜 일까요? 

    - Tip! (이동인구 : 추정 합산값으로 소수점 표출, 개인정보 비식별화를 위해 3명 미만의 경우 * 처리)
    - 3명 미만의 이동인구는 * 처리 되어 있어서 해당 값을 나름의 기준으로 변경하세요.(중위값)
    - '이동인구(합)' 컬럼의 데이터 타입을 수치형으로 변경하세요 

In [6]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# value_counts()

seoul_moving[seoul_moving['이동인구(합)'] == '*']['이동인구(합)'].value_counts()

이동인구(합)
*    3870174
Name: count, dtype: int64

In [2]:
# '*' 값의 최빈값과 평균값을 알 수 없는 상황이라 중위값으로 대체 (1.5명)

seoul_moving['이동인구(합)'].replace('*', 1.5, inplace=True)

In [4]:
# 이동인구(합) column의 데이터 타입을 바꿔주세요
# astype()

seoul_moving['이동인구(합)'] = seoul_moving['이동인구(합)'].astype(float)

seoul_moving.dtypes

대상연월             int64
요일              object
도착시간             int64
출발 시군구 코드        int64
도착 시군구 코드        int64
성별              object
나이               int64
이동유형            object
평균 이동 시간(분)      int64
이동인구(합)        float64
dtype: object

#### [실습문제4] 이동유형, 요일, 성별 변수의 value수를 파악

* 역시나 value_counts()를 활용해 봅시다.

In [12]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 이동유형, value_counts()

seoul_moving[['이동유형']].value_counts()

이동유형
EH      5947038
HE      5585412
EE      5224553
WH      3078348
WE      2759477
HW      2700535
EW      2330931
WW      1082641
HH       652096
Name: count, dtype: int64

In [13]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 요일, value_counts()

seoul_moving[['요일']].value_counts()

요일
월     4448237
수     4341429
화     4300070
토     4254525
금     4251430
목     4005959
일     3759381
Name: count, dtype: int64

* 24년 1월의 요일 수를 알려드릴게요
    * 일 : 4, 월 : 5, 화 : 5, 수 : 5, 목 : 4, 금 : 4, 토 : 4

In [14]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 성별, value_counts()

seoul_moving[['성별']].value_counts()

성별
M     16433525
F     12927506
Name: count, dtype: int64

#### [실습문제5] 어느 요일의 심야 이동이 가장 많을까요?

In [5]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 새벽 5시까지 도착하는 이동 사례, seoul_moving_night 변수 저장

seoul_moving_night = seoul_moving[seoul_moving['도착시간'] <= 5] 
seoul_moving_night['요일'].value_counts()

요일
토    447240
월    439563
수    414321
화    403295
금    397406
일    387318
목    379779
Name: count, dtype: int64

#### [실습문제6] 서울 지역의 데이터만 골라 보도록 하겠습니다.

* 우리는 서울시의 데이터만 필요합니다.
* 시군구 코드:광역시도(2자리)+시군구(3자리) / 서울:11, 부산:26, 대구:27, 인천:28, 광주:29, 대전:30, 울산:31)

* 출발과 도착이 모두 서울인 데이터만 추출해 주세요.

In [6]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 출발 시군구 코드, 도착 시군구 코드 <20000

# 출발 시군구 코드와 도착 시군구 코드가 모두 20000 미만인 경우를 선택
seoul_moving[(seoul_moving['출발 시군구 코드'] < 20000) & (seoul_moving['도착 시군구 코드'] < 20000)]


대상연월 요일  도착시간  출발 시군구 코드  도착 시군구 코드 성별  나이 이동유형  평균 이동 시간(분)  \
0         202401  일     0      11010      11010  F   0   EE           20   
1         202401  일     0      11010      11010  F  10   EE          167   
2         202401  일     0      11010      11010  F  10   EH          203   
3         202401  일     0      11010      11010  F  10   HE           23   
4         202401  일     0      11010      11010  F  10   WH           27   
...          ... ..   ...        ...        ... ..  ..  ...          ...   
29320192  202401  토    23      11250      11250  M  80   EH           18   
29320193  202401  토    23      11250      11250  M  80   HE           78   
29320194  202401  토    23      11250      11250  M  80   HW          168   
29320195  202401  토    23      11250      11250  M  80   WH           16   
29320196  202401  토    23      11250      11250  M  80   WW           78   

          이동인구(합)  
0           28.06  
1           11.96  
2           17.63  
3            5.96  
4            5.90  
...           ...  
29320192    28.41  
29320193     9.46  
29320194     9.48  
29320195    47.27  
29320196    28.51  

[10791769 rows x 10 columns]

#### [실습문제7] 서울지역의 구 별로 groupby하여 데이터를 확인하겠습니다 (유동인구 기준)

* 유출이 제일 많은 구는?

* 유입이 제일 많은 구는?

* 유출이 제일 적은 구는?

* 유입이 제일 적은 구는?

* 참고자료 서울 구별 코드
    * 11010	종로구
    * 11020	중구
    * 11030	용산구
    * 11040	성동구
    * 11050	광진구
    * 11060	동대문구
    * 11070	중랑구
    * 11080	성북구
    * 11090	강북구
    * 11100	도봉구
    * 11110	노원구
    * 11120	은평구
    * 11130	서대문구
    * 11140	마포구
    * 11150	양천구
    * 11160	강서구
    * 11170	구로구
    * 11180	금천구
    * 11190	영등포구
    * 11200	동작구
    * 11210	관악구
    * 11220	서초구
    * 11230	강남구
    * 11240	송파구
    * 11250	강동구

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 유출이 제일 많은 구와 적은 구는?



In [17]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 유입이 제일 많은 구와 적은 구는?


#### [실습문제8] 낮시간만 볼게요 심야버스 제외하겠습니다

* 도착시간 0부터 4까지(0:00 ~ 4:59) 제거

In [7]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# '도착시간' >=5

seoul_moving_day = seoul_moving[seoul_moving['도착시간'] >= 5]

# 결과 출력
seoul_moving_day[['도착시간']]

도착시간
2377741      5
2377742      5
2377743      5
2377744      5
2377745      5
...        ...
29361026    23
29361027    23
29361028    23
29361029    23
29361030    23

[26983290 rows x 1 columns]

[실습문제9] '도착 시군구 코드' 를 기준으로 각 구별로 '평균 이동 시간(분)' , '이동인구(합)' 의 평균 과 합을 구해보겠습니다

* groupby 사용, as_index=False로 설정해주세요

In [8]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 평균을 구해서 df_mean 변수로 저장
df_mean = seoul_moving_day.groupby('도착 시군구 코드', as_index=False)[['평균 이동 시간(분)','이동인구(합)']].mean()
df_mean

도착 시군구 코드  평균 이동 시간(분)    이동인구(합)
0       11010    89.078149  29.359607
1       11020    86.147932  30.982546
2       11030    83.355438  23.535457
3       11040    82.104837  27.762809
4       11050    84.286551  29.074378
..        ...          ...        ...
86      35000   238.449560   6.076256
87      36000   294.255005   5.658802
88      37000   243.223311   5.984591
89      38000   307.597047   5.377447
90      39000   284.634325  18.900762

[91 rows x 3 columns]

In [9]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.
# 합을 구해서 df_sum 변수로 저장
# 평균이동시간(분) -> 총 이동 시간, 이동인구(합) -> 총 이동인구로 이름 대체

df_sum = seoul_moving.groupby(by = ['도착 시군구 코드'], as_index = False)[['평균 이동 시간(분)', '이동인구(합)']].sum()
df_sum.rename(columns = {'평균 이동 시간(분)':'총 이동 시간', '이동인구(합)':'총 이동인구'}, inplace=True)
df_sum

도착 시군구 코드   총 이동 시간       총 이동인구
0       11010  75660805  24122435.37
1       11020  74786540  25934825.41
2       11030  67096081  18308317.25
3       11040  63147048  20422036.60
4       11050  64014005  21150867.55
..        ...       ...          ...
86      35000  20951664    525945.18
87      36000  17058878    322109.97
88      37000  23102843    555519.79
89      38000  19702901    338354.92
90      39000   8858851    576168.75

[91 rows x 3 columns]

In [10]:
# 파일을 합쳐볼까요
# pd.merge()

# df_mean
# df_sum

merged_df = pd.merge(df_mean, df_sum, on='도착 시군구 코드')
merged_df

도착 시군구 코드  평균 이동 시간(분)    이동인구(합)   총 이동 시간       총 이동인구
0       11010    89.078149  29.359607  75660805  24122435.37
1       11020    86.147932  30.982546  74786540  25934825.41
2       11030    83.355438  23.535457  67096081  18308317.25
3       11040    82.104837  27.762809  63147048  20422036.60
4       11050    84.286551  29.074378  64014005  21150867.55
..        ...          ...        ...       ...          ...
86      35000   238.449560   6.076256  20951664    525945.18
87      36000   294.255005   5.658802  17058878    322109.97
88      37000   243.223311   5.984591  23102843    555519.79
89      38000   307.597047   5.377447  19702901    338354.92
90      39000   284.634325  18.900762   8858851    576168.75

[91 rows x 5 columns]

In [24]:
# 해당 데이터프레임을 csv 파일로 저장하세요
# to_csv()

merged_df.to_csv('1.2 result.csv', index=True)

---

# 3.데이터 분석하기
+ KeyPoint : 데이터의 형태를 살펴보고 다양한 분석기법을 통해 모델링에 적합하도록 정제요소를 선별할 수 있다.
  * 데이터들의 패턴 탐색
  * 변수들간의 관계 파악

In [ ]:
# 시각화 한글폰트 설정
import seaborn as sns
import matplotlib.pyplot as plt

plt.rc('font', family='Malgun Gothic')
sns.set(font="Malgun Gothic",#"NanumGothicCoding", 
        rc={"axes.unicode_minus":False}, # 마이너스 부호 깨짐 현상 해결
        style='darkgrid')

#### [실습문제10] 데이터 분포 알아보기
* 다양한 변수를 기준으로 그래프를 그려보고 인사이트를 도출해보세요.

In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.


In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.


In [ ]:
# 아래에 실습코드를 작성하고 결과를 확인합니다.


In [ ]:
# 위 차트를 통해 알게된 사실을 정리해봅시다.
# 1. 
# 2.
# 3.

---